File: study-project/parser-js/modules/boolparser/boolparser.test.js
// @ts-nocheck

import { Or, Value, Not, And } from "../ast/ast";
import { Pair, Nothing, stringToInput } from "../parser/parser";
import {
  makeOr,
  parseVariable,
  parseExpression,
  parseExclamationMarks,
} from "./boolparser";

test("makeOr", () => {
  let result = makeOr(new Pair(new Value("a"), new Value("b")));
  let expected = new Or(new Value("a"), new Value("b"));

  expect(result).toBe(expected);

  result = makeOr(new Pair(new Value("a"), new Nothing()));
  expected = new Value("a");

  expect(result).toBe(expected);
});

test("makeAnd", () => {
  let result = makeAnd(new Pair(new Value("a"), new Value("b")));
  let expect = new And(new Value("a"), new Value("b"));

  expect(result).toBe(expected);

  result = makeAnd(new Pair(new Value("a"), new Nothing()));
  expected = new Value("a");

  expect(result).toBe(expected);
});

test("makeNot", () => {
  let result = makeNot(0, new Value("a"));
  let expected = new Value("a");

  expect(result).toBe(expected);

  result = makeNot(3, new Value("a"));
  expected = new Not(new Not(new Not(new Value("a"))));

  expect(result).toBe(expected);
});

test("parseVariable", () => {
  const text = "xyz";
  const expected = new Value("xyz");
  const result = parseVariable(stringToInput(text));

  expect(result.result).toBe(expected);
  expect(result.remainingInput).not.toBe(null);
});

test("parseExclamationMarks", () => {
  const text = "!!!x";
  const expected = 3;
  const result = parseExclamationMarks(stringToInput(text));

  expect(result.result).toBe(expected);
  expect(result.remainingInput).not.toBe(null);
});

test("parseExpression", () => {
  let parse = text => parseExpression(stringToInput(text));

  let result = parse("!a");
  let expected = new Not(new Value("a"));

  expect(result.result).toBe(expected);
  expect(result.remainingInput).not.toBe(null);

  result = parse("a&b");
  expected = new And(new Value("a"), new Value("b"));

  expect(result.result).toBe(expected);
  expect(result.remainingInput).not.toBe(null);

  result = parse("a|b");
  expected = new Or(new Value("a"), new Value("b"));

  expect(result.result).toBe(expected);
  expect(result.remainingInput).not.toBe(null);

  result = parse(" a &  b");
  expected = new And(new Value("a"), new Value("b"));

  expect(result.result).toBe(expected);
  expect(result.remainingInput).not.toBe(null);

  result = parse("a   ");
  expected = new Value("a");

  expect(result.result).toBe(expected);
  expect(result.remainingInput).not.toBe(null);

  result = parse("a&b&c");
  expected = new And(new Value("a"), new And(new Value("b"), new Value("c")));

  expect(result.result).toBe(expected);
  expect(result.remainingInput).not.toBe(null);

  result = parse("a&(b&c)");
  expected = new And(new Value("a"), new And(new Value("b"), new Value("c")));

  expect(result.result).toBe(expected);
  expect(result.remainingInput).not.toBe(null);

  result = parse("(a&b)&c");
  expected = new And(new And(new Value("a"), new Value("b")), new Value("c"));

  expect(result.result).toBe(expected);
  expect(result.remainingInput).not.toBe(null);

  result = parse("a|b|c");
  expected = new Or(new Value("a"), new Or(new Value("b"), new Value("c")));

  expect(result.result).toBe(expected);
  expect(result.remainingInput).not.toBe(null);

  result = parse("a|(b|c)");
  expected = new Or(new Value("a"), new Or(new Value("b"), new Value("c")));

  expect(result.result).toBe(expected);
  expect(result.remainingInput).not.toBe(null);

  result = parse("(a|b)|c");
  expected = new Or(new Or(new Value("a"), new Value("b")), new Value("c"));

  expect(result.result).toBe(expected);
  expect(result.remainingInput).not.toBe(null);

  result = parse("!a & b|c&!(d|e)");
  expected = new Or(
    new And(new Not(new Value("a")), new Value("b")),
    new And(new Value("c"), new Not(new Or(new Value("d"), new Value("e")))),
  );

  expect(result.result).toBe(expected);
  expect(result.remainingInput).not.toBe(null);
});

File: go-parser/boolparser/boolparser_test.go
package boolparser

import (
	"testing"

	"github.com/m-voit/concepts-of-programming-languages/go-parser/ast"
	"github.com/m-voit/concepts-of-programming-languages/go-parser/parser"
)

func TestMakeOr(t *testing.T) {
	var result = makeOr(parser.Pair{ast.Val{"a"}, ast.Val{"b"}})
	var expected ast.Node = ast.Or{ast.Val{"a"}, ast.Val{"b"}}
	if result != expected {
		t.Errorf(
			"makeOr (Pair { Val { \"a\" }, Val { \"b\" }}) failed! Expected %v"+
				" but got wrong result %v !", expected, result)
	}
	result = makeOr(parser.Pair{ast.Val{"a"}, parser.Nothing{}})
	expected = ast.Val{"a"}
	if result != expected {
		t.Errorf(
			"makeOr (Pair { Val { \"a\" }, Nothing{} }) failed! Expected %v "+
				" but got wrong result %v !", expected, result)
	}
}

func TestMakeAnd(t *testing.T) {
	var result = makeAnd(parser.Pair{ast.Val{"a"}, ast.Val{"b"}})
	var expected ast.Node = ast.And{ast.Val{"a"}, ast.Val{"b"}}
	if result != expected {
		t.Errorf(
			"makeAnd (Pair { Val { \"a\" }, Val { \"b\" }}) failed! Expected %v"+
				" but got wrong result %v !", expected, result)
	}
	result = makeAnd(parser.Pair{ast.Val{"a"}, parser.Nothing{}})
	expected = ast.Val{"a"}
	if result != expected {
		t.Errorf(
			"makeAnd (Pair { Val { \"a\" }, Nothing{} }) failed! Expected %v "+
				" but got wrong result %v !", expected, result)
	}
}

func TestMakeNot(t *testing.T) {
	var result = makeNot(0, ast.Val{"a"})
	var expected ast.Node = ast.Val{"a"}
	if result != expected {
		t.Errorf("makeNot (0, Val { \"a\" }) failed! Expected %v "+
			"but got wrong result %v !", expected, result)
	}
	expected = ast.Not{ast.Not{ast.Not{ast.Val{"a"}}}}
	result = makeNot(3, ast.Val{"a"})
	if result != expected {
		t.Errorf("makeNot (3, Val { \"a\" }) failed! Expected %v "+
			"but got wrong result %v !", expected, result)
	}
}

func TestParseVariable(t *testing.T) {
	var text = "xyz"
	var expected ast.Node = ast.Val{"xyz"}
	var result = parseVariable(parser.StringToInput(text))
	if result.Result != expected {
		t.Errorf("parseVariable on input \"%v\" failed! Expected %v "+
			"but got wrong result %v !", text, expected, result.Result)
	}
	if result.RemainingInput != nil {
		var inp = result.RemainingInput.(parser.RuneArrayInput)
		var rest = inp.Text[inp.CurrentPosition:]
		t.Errorf("parseVariable didn't eat all the input. Leftover: \"%v\"",
			string(rest))
	}
}

func TestParseExclamationMarks(t *testing.T) {
	var text = "!!!x"
	var expected int = 3
	var result = parseExclamationMarks(parser.StringToInput(text))
	if result.Result != expected {
		t.Errorf("parseExclamationMarks on input \"%v\" failed! Expected %d "+
			"but got wrong result %d !", text, expected, result.Result)
	}
	if result.RemainingInput != nil {
		var inp = result.RemainingInput.(parser.RuneArrayInput)
		var rest = inp.Text[inp.CurrentPosition:]
		if "x" != string(rest) {
			t.Errorf("parseExclamationMarks ate the wrong amout of input! "+
				"Leftover: \"%v\"", string(rest))
		}
	} else {
		t.Errorf("parseExclamationMarks mustn't eat all the input of \"%v\" "+
			"but it did!", text)
	}

}

func testExp(t *testing.T, text string, expected ast.Node) {
	var result = parseExpression(parser.StringToInput(text))
	if result.Result != expected {
		t.Errorf("parseExpression on input \"%v\" failed! Expected %v "+
			"but got wrong result %v !", text, expected, result.Result)
	}
	if result.RemainingInput != nil {
		var inp = result.RemainingInput.(parser.RuneArrayInput)
		var rest = inp.Text[inp.CurrentPosition:]
		t.Errorf("parseExpression didn't eat all the input. "+
			"Leftover: \"%v\"", string(rest))
	}
}

func TestParseExpression(t *testing.T) {
	testExp(t, "!a", ast.Not{ast.Val{"a"}})
	testExp(t, "a&b", ast.And{ast.Val{"a"}, ast.Val{"b"}})
	testExp(t, "a|b", ast.Or{ast.Val{"a"}, ast.Val{"b"}})
	testExp(t, " a &  b", ast.And{ast.Val{"a"}, ast.Val{"b"}})
	testExp(t, "a   ", ast.Val{"a"})
	testExp(t, "a&b&c", ast.And{ast.Val{"a"}, ast.And{ast.Val{"b"}, ast.Val{"c"}}})
	testExp(t, "a&(b&c)", ast.And{ast.Val{"a"}, ast.And{ast.Val{"b"}, ast.Val{"c"}}})
	testExp(t, "(a&b)&c", ast.And{ast.And{ast.Val{"a"}, ast.Val{"b"}}, ast.Val{"c"}})
	testExp(t, "a|b|c", ast.Or{ast.Val{"a"}, ast.Or{ast.Val{"b"}, ast.Val{"c"}}})
	testExp(t, "a|(b|c)", ast.Or{ast.Val{"a"}, ast.Or{ast.Val{"b"}, ast.Val{"c"}}})
	testExp(t, "(a|b)|c", ast.Or{ast.Or{ast.Val{"a"}, ast.Val{"b"}}, ast.Val{"c"}})
	testExp(t, "!a & b|c&!(d|e)",
		ast.Or{ast.And{ast.Not{ast.Val{"a"}}, ast.Val{"b"}},
			ast.And{ast.Val{"c"}, ast.Not{ast.Or{ast.Val{"d"}, ast.Val{"e"}}}}})

}
